packages/core/src/sync/index.ts

```diff
@@ -1,318 +1,592 @@
+        };
+      }
+    }
   };
-  blockHash: Hex;
-  blockNumber: Hex;
-  result: null;
-  subtraces: number;
-  traceAddress: number[];
-  type: "reward";
-};
 
-/**
- * Helper function for "eth_getBlockByNumber" request.
- */
-export const _eth_getBlockByNumber = (
-  { requestQueue }: Pick<BaseSyncService, "requestQueue">,
-  {
-    blockNumber,
-    blockTag,
-  }:
-    | { blockNumber: Hex | number; blockTag?: never }
-    | { blockNumber?: never; blockTag: "latest" },
-): Promise<SyncBlock> =>
-  requestQueue
-    .request({
-      method: "eth_getBlockByNumber",
-      params: [
-        typeof blockNumber === "number"
-          ? numberToHex(blockNumber)
-          : blockNumber ?? blockTag,
-        true,
-      ],
-    })
-    .then((_block) => {
-      if (!_block)
-        throw new BlockNotFoundError({
-          blockNumber: (blockNumber ?? blockTag) as any,
+  /**
+   * Estimate optimal range (seconds) to query at a time, eventually
+   * used to determine `to` passed to `getEvents`
+   */
+  let estimateSeconds = 10_000;
+  /**
+   * Omnichain `getEvents`
+   *
+   * Extract all events across `args.networks` ordered by checkpoint.
+   * The generator is "completed" when all event have been extracted
+   * before the minimum finalized checkpoint (supremum).
+   *
+   * Note: `syncStore.getEvents` is used to order between multiple
+   * networks. This approach is not future proof.
+   *
+   * TODO(kyle) programmatically refetch finalized blocks to avoid exiting too early
+   */
+  async function* getEvents() {
+    /**
+     * Calculate checkpoints
+     *
+     * `start`: If `args.initial` is non-zero, use that. Otherwise,
+     * use `start`
+     *
+     * `end`: If every network has an `endBlock` and its less than
+     * `finalized`, use that. Otherwise, use `finalized`
+     */
+    const start =
+      encodeCheckpoint(args.initialCheckpoint) !==
+      encodeCheckpoint(zeroCheckpoint)
+        ? encodeCheckpoint(args.initialCheckpoint)
+        : getChainsCheckpoint("start")!;
+    const end =
+      getChainsCheckpoint("end") !== undefined &&
+      getChainsCheckpoint("end")! < getChainsCheckpoint("finalized")!
+        ? getChainsCheckpoint("end")!
+        : getChainsCheckpoint("finalized")!;
+
+    // Cursor used to track progress.
+    let from = start;
+
+    while (true) {
+      const _localSyncs = args.networks.map(
+        (network) => localSyncs.get(network)!,
+      );
+      // Sync the next interval of each chain.
+      await Promise.all(_localSyncs.map((l) => l.sync()));
+      /**
+       * `latestBlock` is used to calculate the `to` checkpoint, if any
+       * network hasn't yet ingested a block, run another iteration of this loop.
+       * It is an invariant that `latestBlock` will eventually be defined. See the
+       * implementation of `LocalSync.latestBlock` for more detail.
+       */
+      if (_localSyncs.some((l) => l.latestBlock === undefined)) continue;
```

Comment:
Reading this line, the naming of `localSync.latestBlock` is a bit confusing considering this is the historical sync (would expect to see "historical" somewhere).